Split RHEL 8 CIS profile using new controls file format

alexhaydock opened the pull request:
#### Description:

The current CIS profile for RHEL 8 is a monolithic single-profile combining the CIS rules. This does not address the fact that the CIS benchmark is published by CIS as 4 overlapping but separate sub-profiles:

- Level 1 - Server
- Level 1 - Workstation
- Level 2 - Server
- Level 2 - Workstation

This PR is an attempt to address this (for RHEL 8 initially, but I will expand this to RHEL 7 and 6 later). I intend to also do the same for CentOS and Rocky eventually.

#### Rationale:

This split uses the current profile format and simply splits it into 4 separate profiles which acknowledge the different rules which are enabled/disabled in each.

More information about this rationale can be found in my discussion thread #6924.

I considered an approach involving the more modular approach that the ANSSI profile is now taking, but since the CIS benchmarks are not cleanly "hierarchical" this doesn't work as cleanly here. At the very least we'd still need separate Server and Workstation profiles.

I know there might still be more work needed before this can be merged, and I'm also aware that there's a number of outstanding controls here which need rules adding. I can see that the upcoming RHEL 8.4 beta has rebased their CAC content to 0.1.54.

My (possibly ambitious) goal here is to land this change (or a similar one), and attempt to clean up any missing CIS rules in time for RHEL 8.5 to include a complete set of CIS profiles when it releases in late 2021/early 2022.

I would especially welcome any comments from @cipherboy or @vojtapolasek on this one, as I know they were involved in some of the discussions I had when planning for this. (No pressure though, I know big monolithic changes like this one can't be fun to review).

Thanks all :smile:

cipherboy commented:
/ok-to-test

alexhaydock commented:
I'm getting some errors in the CI jobs related to this change like:

```
 The following tests FAILED:
	 69 - missing-references-ssg-rhel8-xccdf.xml (Failed)
```

and

```
The following tests FAILED:
	142 - missing-references-ssg-rhel8-xccdf.xml (Failed)
```

I've pushed a commit which updates the RHEL8 `CMakeLists.txt` file to hopefully address this but I might be missing some more changes that need making before the new profiles are referenced in all the right places.

cipherboy commented on rhel8/CMakeLists.txt:
```diff
@@ -23,6 +22,11 @@ ssg_build_html_anssirefs_table(${PRODUCT} "bp28_intermediary")
 ssg_build_html_anssirefs_table(${PRODUCT} "bp28_enhanced")
 ssg_build_html_anssirefs_table(${PRODUCT} "bp28_high")
 
+ssg_build_html_cisrefs_table(${PRODUCT} "cis_l1_server")
```
Up to the RHEL folks, but I'd maybe consider spelling out `level` to make it consistent with the proposed Ubuntu #6416 PR?

alexhaydock replied:
I couldn't really decide either way when choosing the naming, but this seems sensible enough and I think there's a definite benefit to consistency. I've added a commit to switch to "level".

alexhaydock commented:
/retest

JAORMX commented:
@openscap-ci test this please

alexhaydock commented:
Pushed a new commit to fix the build issue the CI job was uncovering. I'd added a new HTML table type for CIS to the RHEL 8 `CMakeLists.txt` to be consistent with the other benchmarks with "sub-profiles" like ANSSI and NIST, but after looking at how these are defined in `SSGCommon.cmake`, I'm not sure we need a specific one for CIS, so instead I'm just using the regular `ssg_build_html_table_by_ref` once for each CIS sub-profile.

alexhaydock commented:
At this point I'm getting a bit stuck getting this PR in order. I'm not sure my last commit at 24872ebd7e110063e00376d1f9db1cf430726294 actually fixes things or whether it's a band-aid over the CI failure I was getting before.

I've tried to read the developer docs to get a better understanding of how the HTML tables are generated (which seems to be what's failing currently), but I'm struggling to see the process that's going on.

I also don't want to patch things up by just copying things 4 times for these CIS profiles, when there could be a single unified table schema referred to once.

If anyone's able to help point me in the right direction, that would be much appreciated. :smile:

alexhaydock commented:
Something is going wrong with the table build process with the current commit.

An example is below, that gets built as `tables/table-rhel8-cisrefs-level1_server.html`

```html
<html xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:cdf="http://checklists.nist.gov/xccdf/1.1" xmlns:xhtml="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title></title>
</head>
<body>
<br><br><div style="text-align: center; font-size: x-large; font-weight:bold"></div>
<br><br><style type="text/css">
		table
		{
			border-collapse:collapse;
		}
		table, th, td
		{
			border: 2px solid #dcdcdc;
			border-left: none;
			border-right: none;
			vertical-align: top;
			padding: 2px;
			font-family: verdana,arial,sans-serif;
			font-size:11px;
		}
		pre { 
			white-space: pre-wrap;
			white-space: -moz-pre-wrap !important;
			word-wrap:break-word; 
		}
		table tr:nth-child(2n+2) { background-color: #f4f4f4; }
		thead
		{
			display: table-header-group;
			font-weight: bold;
			background-color: #dedede;
		}
	</style>
<table><thead>
<td>CCE ID</td>
<td>Rule Title</td>
<td>Description</td>
<td>Rationale</td>
<td>Variable Setting</td>
<td>CIS v1.1.0  Mapping</td>
</thead></table>
</body>
</html>
```

Clearly something isn't being referenced right here in the table generation process. It's not being populated with the right content and the test is failing as a result.

vojtapolasek commented:
I suggest to wait for introduction of tags into the policy definition files.

alexhaydock commented:
> I suggest to wait for introduction of tags into the policy definition files.

What would be the next steps to work towards this do you think? I think some of the discussion in #6996 is now focusing on aspects which might be more CIS-specific and might be more relevant upstream (which is great! but might not help us in the shorter-term).

ggbecker commented on rhel8/profiles/cis_level2_server.profile:
```diff
@@ -8,12 +8,14 @@ metadata:
 
```
I just wanted to point out that renaming this file means renaming a profile ID and this breaks compatibility with previous versions and tailoring, in other words if people has automated scanning that uses old 'cis' profile, it will break because the profile won't exist anymore in newer versions. We should probably keep the way it is and may add `cis_level2_server.profile` that will extend the `cis` profile, but this is also not ideal.

alexhaydock replied:
That makes sense - thanks for taking a look :smile: 

I was taking some inspiration from the Ubuntu PR at #6416, but after the discussion going on in #6996, I think the conclusion might be that we wait for tags to be available in `.policy` files and then I can rework this PR to just make edits to the original `cis.profile` instead of splitting it out.

vojtapolasek replied:
I agree with @ggbecker . Currently, the profile contains probably mix of server and workstation rules, both level 1 and level 2. We have to keep this profile ID and I would suggest to use "cis" ID for "Level 2 server".

alexhaydock replied:
I agree. I think if we have to maintain the original ID to avoid breaking people's scripts, "Level 2 - Server" makes sense to be the one to pick. If they're using the current profile then they're already applying or auditing against the Level 2 controls so this approach avoids unexpectedly switching people to a more lenient profile if we started using "cis" to refer to Level 1. We can use other IDs for the other 3 profiles.

vojtapolasek commented:
@alexhaydock This PR introduces changes to the way in which policy levels are defined.
https://github.com/ComplianceAsCode/content/pull/7072
Do you plan to use it while splitting rhel8 profile?
I am going to use it while splitting the rhel7 profile.

alexhaydock commented:
> @alexhaydock This PR introduces changes to the way in which policy levels are defined.
> #7072
> Do you plan to use it while splitting rhel8 profile?
> I am going to use it while splitting the rhel7 profile.

@vojtapolasek That's my plan. I think overall this PR will need a complete rework, but I think we're in a much better situation now that #7072 has been merged. I'll try and put aside some time to work on this soon. In the meantime I've been trying to increase the coverage of some of the CIS rules that are missing.

alexhaydock commented:
I've brought this branch back in line with `master` as rebasing didn't really make sense with how different this is going to end up now that #7108 is merged. I'm still planning to fix this one but GitHub has auto-closed the PR for now. Stay tuned.

alexhaydock commented:
@vojtapolasek - A quick question about how we're approaching "Automated" vs "Manual" rules here --

I'm working on splitting these out currently and I've noticed when hitting a "manual" control that is present in both RHEL 7 and RHEL 8 that you've opted not to do anything for that control, but previously in the RHEL 8 CIS profile it's a control we have been remediating.

This is an example of what I mean. It's just the first one I hit of this type:

In the RHEL 7 control file:
https://github.com/ComplianceAsCode/content/blob/5ba05c306b39273048ec08732bdeb8d5dc64b6ee/controls/cis_rhel7.yml#L301-L306

By my understanding, OpenSCAP won't take any action when "applying" this control. But I notice that in our current RHEL 8 profile, we are in fact applying something for this control:
https://github.com/ComplianceAsCode/content/blob/5ba05c306b39273048ec08732bdeb8d5dc64b6ee/products/rhel8/profiles/cis.profile#L114-L115

Is the general policy that we shouldn't be automatically applying controls which are listed as "Manual"? In that case I'm wondering what the function of the section you have above in the RHEL 7 control file is. It seems that if there's no rule referenced then OpenSCAP will skip over this control and we can't even use it for auditing. If there's an existing rule that matches a "Manual" control, should we be referencing it and setting `automated: no`? Or are we using these rules effectively as documentation, to show that every rule from the CIS Benchmarks has been addressed in the control file, even if some of them don't actually do anything?

vojtapolasek commented:
Hello, great question.
> @vojtapolasek - A quick question about how we're approaching "Automated" vs "Manual" rules here --
> 
> I'm working on splitting these out currently and I've noticed when hitting a "manual" control that is present in both RHEL 7 and RHEL 8 that you've opted not to do anything for that control, but previously in the RHEL 8 CIS profile it's a control we have been remediating.
> 
> This is an example of what I mean. It's just the first one I hit of this type:
> 
> In the RHEL 7 control file:
> 
> https://github.com/ComplianceAsCode/content/blob/5ba05c306b39273048ec08732bdeb8d5dc64b6ee/controls/cis_rhel7.yml#L301-L306
> 
> By my understanding, OpenSCAP won't take any action when "applying" this control. But I notice that in our current RHEL 8 profile, we are in fact applying something for this control:
> 
> https://github.com/ComplianceAsCode/content/blob/5ba05c306b39273048ec08732bdeb8d5dc64b6ee/products/rhel8/profiles/cis.profile#L114-L115
> 
> Is the general policy that we shouldn't be automatically applying controls which are listed as "Manual"?
Yes I took this approach when doing the RHEL 7 split. In the past, CIS controls were marked as "scored" and "not scored". My understanding was that this does not say if the control should be automated or not. Later, CIS changed benchmarks and now there are "automated" and "manual" controls. The CIS benchmark says that manual control

```
Represents recommendations for which assessment of a technical control cannot be fully automated and requires all or some manual steps to validate that the configured state is set as expected. 
```

I admit that for some cases we might have rules which might partially automate even controls which are marked as "manual". But there are not many of them. I suggest that we keep "manual" controls without automation. My goal is to achieve "not wrong" profile. That means maybe use less rules, but make sure that they are not getting in the way of administrators. So if some "manual" control for example allows an administrator to choose what NTP server they want to use, I would leave it to the administrator.

> In that case I'm wondering what the function of the section you have above in the RHEL 7 control file is. It seems that if there's > no rule referenced then OpenSCAP will skip over this control and we can't even use it for auditing. If there's an existing rule that > matches a "Manual" control, should we be referencing it and setting `automated: no`? Or are we using these rules effectively as > documentation, to show that every rule from the CIS Benchmarks has been addressed in the control file, even if some of them don't actually do anything?

My intention is to have all controls in the policy file, even those which are not automated. It allows to easily review future changes to the benchmark. If you want, you can use the "related_rules" key to add suggested rules to the "manual" controls. But this won't be visible in the resulting datastream, it brings value to the policy file authors.

alexhaydock commented:
> My intention is to have all controls in the policy file, even those which are not automated. It allows to easily review future changes to the benchmark. If you want, you can use the "related_rules" key to add suggested rules to the "manual" controls. But this won't be visible in the resulting datastream, it brings value to the policy file authors.

So for the above example we could do:
```yaml
 - id: 1.2.5 
   title: Disable the rhnsd Daemon (Manual) 
   levels: 
     - l1_server 
     - l1_workstation 
   automated: no
   related_rules:
     - service_rhnsd_disabled
```

Does that make sense? Are the `related_rules` purely for policy authors reading the policy files here, or do they show up in tools like SCAP Workbench?

vojtapolasek commented:
> > My intention is to have all controls in the policy file, even those which are not automated. It allows to easily review future changes to the benchmark. If you want, you can use the "related_rules" key to add suggested rules to the "manual" controls. But this won't be visible in the resulting datastream, it brings value to the policy file authors.
> 
> So for the above example we could do:
> 
> ```yaml
>  - id: 1.2.5 
>    title: Disable the rhnsd Daemon (Manual) 
>    levels: 
>      - l1_server 
>      - l1_workstation 
>    automated: no
>    related_rules:
>      - service_rhnsd_disabled
> ```
> 
Yes, I think this is fine. However, I did not do this consistently in rhel7 split.

> Does that make sense? Are the `related_rules` purely for policy authors reading the policy files here, or do they show up in tools like SCAP Workbench?

Currently they are only for policy authors.

alexhaydock commented:
Making progress on this, but it's slow progress. I'm working on compiling the control file first, and then I'll work on making sure all of the CIS references in the individual rule files are up-to-date for the `1.0.1` benchmark release. I'm guessing there'll be a lot less change here than with the RHEL 7 one since `1.0.0` to `1.0.1` isn't likely to carry any big changes.

vojtapolasek commented:
No problem, it took me over 5 days to go through rhel7 CIS. Maybe we could work in parallel? You could propose a PR which just sets the basics - creates the policy file and respective profiles. Then we could somehow split the benchmark (per section?) and propose PRs.

alexhaydock commented:
@vojtapolasek From looking at the appendix of changes in the back of the `1.0.1` RHEL benchmark, there might not actually be much to do here other than creating the policy file and respective profiles. I don't think many (if any) of the current rules will need their references updating for RHEL 8.

There's still some outstanding rules that will need some thought. We can definitely split those up and tackle them if you like. I was going to go for the low-hanging fruit first (i.e. the ones that CIS-CAT still reports as failures on a system which has been remediated with the current profile).

There's some that will definitely need some deeper thought though since there'll have to be a lot of conditional logic involved. The most obvious example to me is all of the controls that apply to ntables, or the ones that apply to iptables. Those controls must only apply if nftables or iptables is already the chosen firewall service on the system, and presumably there needs to be some logic to work that out. And then we need to think about what to do when, for example, firewalld and nftables are both enabled. A lot of those rules are marked as automated. We should really do _something_ since under CIS it's perfectly valid to replace firewalld with nftables, or with iptables, so I don't think we should be opinionated like we currently are and just enforce firewalld, even if that probably works for 99% of end users.

alexhaydock commented:
@vojtapolasek I think I'm at a point where this could possibly be reviewed for merge.

---

### What has been done
- The `cis.profile` has been reworked to reference the control file, so that `cis` refers to "Level 2 - Server"
- Profile files have been created for the other 3 CIS benchmark levels, which also refer to the new control file
- Every control from the `1.0.1` RHEL 8 benchmark has a corresponding entry in the control file
- All control titles match the ones in the up-to-date `1.0.1` benchmark
- `TODO` and `NEEDS RULE` comments have been preserved from the original `cis.profile` file for all rules marked "Automated"
- I have commented on and closed (thanks @yuumasato!) all the issues I could find where they were present as a `NEEDS RULE` comment but were for rules now marked "Manual" (i.e. where we don't actually need the rule)
- I have opened new `NEEDS RULE` issues for all of the Automated rules which didn't already have an issue, and added links to the issues to comments in the control file

### What hasn't been done
- I moved the pre-existing `NEEDS RULE` comments across without checking whether any relevant rules have been created since the comments were added. It's possible that we have some of these `NEEDS RULE` comments which rules have since been made for, but we will have to check.
- I haven't updated the CIS references for RHEL 8 in the rule files themselves. I spoke too soon in my comment earlier as I found some examples of this later on, but I _think_ they're all confined to Section 6, where a bunch of the rule numbers got swapped around. I think that's probably easiest to do in a separate PR though.

vojtapolasek commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2268 @@
+      - l1_workstation
+    automated: yes
+    rules:
+      - configure_crypto_policy
```
The rule is missing a cis reference.
Additionally, I suggest we add

```
      - var_system_crypto_policy=default
```

alexhaydock replied:
I'm not sure I can find where this rule is missing the CIS reference. It seems to have one for RHEL 8 but not RHEL 7:
https://github.com/ComplianceAsCode/content/blob/2b2152d288e05f0d64f26fff3f01b0e75311023d/linux_os/guide/system/software/integrity/crypto/configure_crypto_policy/rule.yml#L64

Additionally, if we set `var_system_crypto_policy` to `default` here, how does this interact with what I'm doing for CIS 1.11 below? If two variables are set in a controls file, which one ends up taking precedence?

vojtapolasek replied:
You are right, I overlooked it cause the rule has actually two references. And they are correct.

vojtapolasek reviewed (changes requested):
Hi @alexhaydock  and thank you very much. I reviewed the section one, please see comments.
I will continue reviewing other sections tomorrow. It is great that you create issues for missing rules.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2268 @@
+    automated: yes
+    rules:
+      - file_permissions_etc_issue
+
```
This one was an oversight. I've added this too and also filed #7225 as an issue for the missing rule.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2268 @@
+      - l1_server
+      - l1_workstation
+    automated: no
+    rules:
```
Fixed.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2268 @@
+      - l1_server
+      - l1_workstation
+    automated: no
+    rules:
```
Fixed.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2268 @@
+      - l1_server
+      - l1_workstation
+    automated: no
+    rules:
```
Fixed.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2268 @@
+    rules:
+      - package_libselinux_installed
+
+  - id: 1.7.1.1
```
Done.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2268 @@
+      - l1_workstation
+    automated: yes
+    rules:
+      - banner_etc_issue
```
Added as above.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2268 @@
+    rules:
+      - dconf_gnome_banner_enabled
+      - dconf_gnome_login_banner_text
+
```
Added, thanks.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2268 @@
+      - file_groupowner_grub2_cfg
+      - file_permissions_grub2_cfg
+
+  - id: 1.5.1
```
Fixed.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2268 @@
+      - l1_workstation
+    automated: yes
+    rules:
+      - grub2_password
```
This makes sense. Added.

alexhaydock reviewed (commented):
Thanks @vojtapolasek. I've resolved most of the issues here and pushed a new commit. Some might need some more thought, like the one about setting permissions for GRUB2 EFI that doesn't necessarily align directly with the benchmark. I'll take a closer look at those soon.

vojtapolasek commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2285 @@
+      - l1_workstation
+    automated: yes
+    rules:
+      - configure_crypto_policy
```
After this PR gets merged, you can use the default policy for this rule and future policy for the next rule:
https://github.com/ComplianceAsCode/content/pull/7226

alexhaydock replied:
Ah perfect! - I'll add this in now and we can just block merging of this PR until after #7226 gets merged.

vojtapolasek commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2285 @@
+    related_rules:
+      - package_chrony_installed
+
+  - id: 2.1.1
```
```suggestion
  - id: 2.2.1.2
```

alexhaydock replied:
Fixed

vojtapolasek commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2285 @@
+      - l1_workstation
+    automated: yes
+    rules:
+      - service_chronyd_enabled
```
Why this rule? It is not mentioned in the benchmark.
Moreover, I suggest adding this variable:
var_multiple_time_servers=rhel

alexhaydock replied:
This is what we're using in the current profile.

https://github.com/ComplianceAsCode/content/blob/21d5188001edba157e5d925e0012a5609d77f00b/products/rhel8/profiles/cis.profile#L271-L274

This is an interesting one though. I agree that, by pure wording, the benchmark doesn't actually call for Chrony to be running, but between 2.2.1.1 calling for Chrony to be _installed_, and this control calling for Chrony to be _configured_, I feel like it's implied that the goal of the benchmark is for the service should actually be running and enabled.

But if we want to be strictly correct as per the benchmark then it doesn't call for this, so we could remove this rule.

vojtapolasek commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2285 @@
+      - chronyd_run_as_chrony_user
+
+  - id: 2.2.2
+    title: Ensure chrony is configured (Automated)
```
```suggestion
    title: Ensure X Window System is not installed (Automated)
```

alexhaydock replied:
Fixed

vojtapolasek commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2285 @@
+      - l1_server
+    automated: yes
+    rules:
+      - package_xorg-x11-server-common_removed
```
I suggest replacing both rules with this one, it performs removel of more packages and configures the correct run leve:
xwindows_remove_packages

alexhaydock replied:
After looking at this, I agree

vojtapolasek commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2285 @@
+      - l1_workstation
+    automated: yes
+    rules:
+      - package_squid_removed
```
I think service_squid_disabled is better aligned with the benchmark.

alexhaydock replied:
That seems fair - fixed.

vojtapolasek commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2285 @@
+  # NEEDS RULE
+  # https://github.com/ComplianceAsCode/content/issues/5231
+  - id: 2.2.14
+    title: Ensure RPC is not enabled (Automated)
```
```suggestion
    title: Ensure LDAP server is not enabled (Automated)
```

alexhaydock replied:
Whoops - fixed :smile:

vojtapolasek commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2285 @@
+      - l1_workstation
+    automated: yes
+    rules:
+      - postfix_network_listening_disabled
```
I would also add this variable:
var_postfix_inet_interfaces=loopback-only

alexhaydock replied:
Good catch - added.

vojtapolasek reviewed (changes requested):
REviewed section 2. See comments.

vojtapolasek reviewed (changes requested):
Reviewed section 3.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2285 @@
+    automated: yes
+    rules:
+      - sysctl_net_ipv4_conf_all_accept_source_route
+      - sysctl_net_ipv4_conf_default_accept_source_route
```
Added

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2285 @@
+    rules:
+      - sysctl_net_ipv4_conf_all_accept_redirects
+      - sysctl_net_ipv4_conf_default_accept_redirects
+      - sysctl_net_ipv6_conf_all_accept_redirects
```
It looks like all 4 of these have CIS references.

These are correct for both RHEL 7 and RHEL 8:

https://github.com/ComplianceAsCode/content/blob/21d5188001edba157e5d925e0012a5609d77f00b/linux_os/guide/system/network/network-kernel/network_host_and_router_parameters/sysctl_net_ipv4_conf_all_accept_redirects/rule.yml#L32-L33

https://github.com/ComplianceAsCode/content/blob/21d5188001edba157e5d925e0012a5609d77f00b/linux_os/guide/system/network/network-ipv6/configuring_ipv6/sysctl_net_ipv6_conf_default_accept_redirects/rule.yml#L24-L25

This one is wrong for RHEL 7:

https://github.com/ComplianceAsCode/content/blob/21d5188001edba157e5d925e0012a5609d77f00b/linux_os/guide/system/network/network-kernel/network_host_and_router_parameters/sysctl_net_ipv4_conf_default_accept_redirects/rule.yml#L31-L32

And this one is wrong for RHEL 8:

https://github.com/ComplianceAsCode/content/blob/21d5188001edba157e5d925e0012a5609d77f00b/linux_os/guide/system/network/network-ipv6/configuring_ipv6/sysctl_net_ipv6_conf_all_accept_redirects/rule.yml#L24-L25

I've also noticed that you're not applying the "default" component of this rule for RHEL 7. I will submit a PR later correcting this as well as the above:

https://github.com/ComplianceAsCode/content/blob/21d5188001edba157e5d925e0012a5609d77f00b/controls/cis_rhel7.yml#L876-L886

alexhaydock replied:
Created #7259 to address this.

vojtapolasek replied:
Aha, then I am sorry, maybe I overlooked the reference. Thanks for the PR, I will look at it.

alexhaydock reviewed (commented):
Addressed comments for Section 3

vojtapolasek commented on controls/cis_rhel8.yml:
```diff
@@ -638,7 +638,7 @@ controls:
     automated: yes
     rules:
-      - package_squid_removed
+      - package_squid_disabled
```
```suggestion
      - service_squid_disabled
```

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2268 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: yes
```
This makes sense - I've set this now.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2268 @@
+      - file_owner_grub2_cfg
+      - file_groupowner_grub2_cfg
+      - file_permissions_grub2_cfg
+
```
The benchmark does take an interesting approach here, setting the `umask` and `fmask` of the `/boot/efi` mountpoint in `/etc/fstab` rather than merely changing the permissions of the files, but I think the end-result is effectively the same and our actions here will satisfy the audit and rationale from the benchmark even if it doesn't use the exact pattern laid out in the remediation section for this control.

With that in mind, I've added the rules you suggest - which also brings RHEL 8's version of this rule into line with how you've done it for RHEL 7.

alexhaydock reviewed (commented):
Hi @vojtapolasek 

Nice to see a new CAC release out! It prompted me to go back through this PR and make sure I've addressed all of the comments so far. I think I've addressed everything now in the PR I just pushed except the Chrony one, which you might want to take another look at.

vojtapolasek reviewed (changes requested):
Reviewed section 4.

vojtapolasek commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: yes
```
```suggestion
    automated: partially
```

alexhaydock replied:
Noted

vojtapolasek commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: no
```
I think we have rule for part of this requirement: use_pam_wheel_for_su
```suggestion
    automated: partially
```

alexhaydock replied:
Added

vojtapolasek commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+      - l1_server
+      - l1_workstation
+    automated: yes
+    rules:
```
How about this one as well?
accounts_umask_etc_login_defs

alexhaydock replied:
Added

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+      - l2_workstation
+    automated: yes
+    rules:
+      - auditd_data_retention_max_log_file
```
I agree. I've set this to 6, since that seems to be the value used by a lot of other profiles, and it seems to be the default. I personally keep a lot more on my systems but not changing this value will help avoid breaking anyone's deployment pipelines if they're not creating very large `/var/log/audit` volumes.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+      - l2_workstation
+    automated: yes
+    rules:
+      - auditd_data_retention_max_log_file_action
```
Done

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    rules:
+      - audit_rules_time_adjtimex
+      - audit_rules_time_clock_settime
+      - audit_rules_time_settimeofday
```
I've set this to partial and added a note explaining this fact

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l2_server
+      - l2_workstation
+    automated: yes
```
Fixed

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+      - audit_rules_unsuccessful_file_modification_openat
+      - audit_rules_unsuccessful_file_modification_truncate
+      # Opinionated selection
+      - audit_rules_unsuccessful_file_modification_open_by_handle_at
```
Good question - This came directly from the previous profile. Since it's not mentioned in the benchmark, I've removed it.

vojtapolasek commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: yes
```
This rule lacks checks and remediations. I would drop it or move it into related_rules.
```suggestion
    automated: no
```

alexhaydock replied:
Done

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+      - l2_workstation
+    automated: yes
+    rules:
+      - audit_rules_privileged_commands
```
I've set it to `automated: no` and added a note to this effect.

vojtapolasek reviewed (changes requested):
REviewing section 6. Thank you very much for this! Great job!

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+      - audit_rules_file_deletion_events_unlink
+      - audit_rules_file_deletion_events_unlinkat
+      # Opinionated selection
+      - audit_rules_file_deletion_events_rmdir
```
This one is the same story as above, so I've removed it.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+      - l1_workstation
+    automated: yes
+    rules:
+      - rsyslog_files_permissions
```
Done, and I've created #7332 to track this one.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+      - l1_workstation
+    automated: yes
+    rules:
+      - rsyslog_remote_loghost
```
Sorted - and I've opened #7333 to track this one.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+      - l1_workstation
+    automated: yes
+    rules:
+      - file_permissions_sshd_private_key
```
Addressed with a comment and I have opened #7334 to track.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: yes
```
Addressed

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+      - l1_workstation
+    automated: yes
+    rules:
+      - sshd_idle_timeout_value=5_minutes
```
I'm not quite sure what this one changed?

vojtapolasek replied:
eeech sorry, there should be 15 minutes

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    rules:
+      - sshd_idle_timeout_value=5_minutes
+      - sshd_set_idle_timeout
+      - sshd_set_keepalive_0
```
Ah this makes sense, I've kept the variable in in case a user wants to override it with a tailoring file.

alexhaydock reviewed (commented):
Still a work in progress reviewing these - thanks for the comments

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+      - l1_workstation
+    automated: yes
+    rules:
+      - sshd_set_maxstartups
```
Done

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    automated: yes
+    rules:
+      - sshd_set_max_sessions
+      - var_sshd_max_sessions=4
```
I've changed this to your suggestion here so that we match the audit check for this rule but I'm _very_ confused where CIS got the title of this one from, since it does very clearly say `4`... 😕

alexhaydock replied:
I filed an upstream issue asking about this one. https://workbench.cisecurity.org/community/14/tickets/13414

alexhaydock replied:
Resolved using additional info from upstream and Draft Benchmark 1.1.0

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+
+  # NEEDS RULE
+  # https://github.com/ComplianceAsCode/content/issues/5532
+  - id: 5.3.2
```
Fixed

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: yes
```
I removed these and filed #7337 to track.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+      - var_accounts_passwords_pam_faillock_unlock_time=900
+
+  - id: 5.4.3
+    title: Ensure password reuse is limited (Automated)
```
I've brought over the rules and notes from the RHEL 7 controls file here.

This `notes:` field is new to me. So far I've been placing my notes inline as comments around the controls as I'm sure you've seen. Does the `notes:` field get rendered anywhere user-visible or is it just for upstream when working on the control file itself?

vojtapolasek replied:
Notes are rendered only upstream, they are not conveyed to the end user.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: yes
```
Updated with note

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: yes
```
Updated with note

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: yes
```
Updated with note

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: yes
```
Updated with note

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: yes
```
Updated with note

alexhaydock reviewed (commented):
Addressed some more review comments - have almost completed reviewing Section 5.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: yes
```
Amended with note, and #7352 opened

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: yes
```
Noted

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: no
```
Looking at the OVALs for these two rules, I think it will satisfy the benchmark to use them.

- `accounts_root_path_dirs_no_write` checks for group/world writable files in the root $PATH
- `root_path_no_dot` checks for `.` in root's $PATH, but also checks for the other things that the benchmark mandates (empty dirs & trailing `:` characters)

I have added these.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: yes
```
Done

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    levels:
+      - l1_server
+      - l1_workstation
+    automated: yes
```
Done

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+    automated: no
+
+  - id: 6.2.20
+    title: Ensure shadow group is empty (Automated)
```
Fixed

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+      - l2_workstation
+    automated: yes
+    rules:
+      - audit_rules_kernel_module_loading
```
Added, and I have validated the CIS RHEL 8 references for these rules and opened #7353 to update some of them.

alexhaydock reviewed (commented):
Thanks again for the in-depth review, @vojtapolasek. I think I've addressed everything except the mixed-up identifiers in section 6, which I'll do now as part of a separate PR.

alexhaydock commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2300 @@
+      - l1_server
+      - l1_workstation
+    automated: yes
+    rules:
```
I've checked all the rules we're using in 6.1 and opened #7356 to sort these out.

alexhaydock commented:
@vojtapolasek

I think this is in good condition now with everything reviewed.

If you're happy with the state of it, I can try a rebase against master to get it ready.

We should also aim to merge PRs #7356 and #7353 along with this one so everything is in sync.

alexhaydock commented:
One more change pushed - after confirming upstream about the discrepancy between the wording of the title and the audit process for "5.2.18 Ensure SSH MaxSessions is set to 4 or less", I have confirmed that the intended value for this control is definitely `4`, and not `10` as can be seen in the Audit section for the control in benchmark 1.0.1

I'm resetting this to 4 to match the spirit of the benchmark, and to align with both previous versions and the forthcoming 1.1.0 release.

https://github.com/alexhaydock/ComplianceAsCode-content/blob/b0807115a757fc85aaeb59e3b5a69f21a2eb527d/controls/cis_rhel8.yml#L1784-L1799

vojtapolasek commented:
I agree, it looks good. I think we also need this PR merged so that selinux_policy is correctly set in l1 and l2; there are different options for those levels and currently it does not work correctly.
https://github.com/ComplianceAsCode/content/pull/7226

vojtapolasek commented:
But when I think about that, I think you can do the rebase already. The PR mentioned above should not cause additional rebases.

pep8speaks commented:
Hello @alexhaydock! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`build-scripts/profile_tool.py`](https://github.com/ComplianceAsCode/content/blob/150e581caefa6a6ec81375f01018c617c5ef1f16/build-scripts/profile_tool.py):

> [Line 48:25](https://github.com/ComplianceAsCode/content/blob/150e581caefa6a6ec81375f01018c617c5ef1f16/build-scripts/profile_tool.py#L48): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 49:25](https://github.com/ComplianceAsCode/content/blob/150e581caefa6a6ec81375f01018c617c5ef1f16/build-scripts/profile_tool.py#L49): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 72:25](https://github.com/ComplianceAsCode/content/blob/150e581caefa6a6ec81375f01018c617c5ef1f16/build-scripts/profile_tool.py#L72): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 73:25](https://github.com/ComplianceAsCode/content/blob/150e581caefa6a6ec81375f01018c617c5ef1f16/build-scripts/profile_tool.py#L73): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent







* In the file [`shared/templates/audit_rules_dac_modification/template.py`](https://github.com/ComplianceAsCode/content/blob/150e581caefa6a6ec81375f01018c617c5ef1f16/shared/templates/audit_rules_dac_modification/template.py):

> [Line 5:100](https://github.com/ComplianceAsCode/content/blob/150e581caefa6a6ec81375f01018c617c5ef1f16/shared/templates/audit_rules_dac_modification/template.py#L5): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (114 > 99 characters)





















* In the file [`tests/install_vm.py`](https://github.com/ComplianceAsCode/content/blob/150e581caefa6a6ec81375f01018c617c5ef1f16/tests/install_vm.py):

> [Line 198:100](https://github.com/ComplianceAsCode/content/blob/150e581caefa6a6ec81375f01018c617c5ef1f16/tests/install_vm.py#L198): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (395 > 99 characters)









* In the file [`utils/create-stig-overlay.py`](https://github.com/ComplianceAsCode/content/blob/150e581caefa6a6ec81375f01018c617c5ef1f16/utils/create-stig-overlay.py):

> [Line 146:23](https://github.com/ComplianceAsCode/content/blob/150e581caefa6a6ec81375f01018c617c5ef1f16/utils/create-stig-overlay.py#L146): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 146:100](https://github.com/ComplianceAsCode/content/blob/150e581caefa6a6ec81375f01018c617c5ef1f16/utils/create-stig-overlay.py#L146): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 99 characters)
> [Line 160:17](https://github.com/ComplianceAsCode/content/blob/150e581caefa6a6ec81375f01018c617c5ef1f16/utils/create-stig-overlay.py#L160): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 177:100](https://github.com/ComplianceAsCode/content/blob/150e581caefa6a6ec81375f01018c617c5ef1f16/utils/create-stig-overlay.py#L177): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 99 characters)

alexhaydock commented:
@vojtapolasek It took a few tries to work out, but I think I got the rebase to work properly and it doesn't seem to have caused any issues. My branch is up to date with master now.

vojtapolasek commented on controls/cis_rhel8.yml:
```diff
@@ -0,0 +1,2324 @@
+    levels:
+      - l2_server
+      - l2_workstation
+    automated: yes
```
Hello, I incidentally stumbled upon this. This needs to be fixed. If the rule is marked as manual, we do not automate it. Moreover, this rule no longer works in rhel7 and later, so I wouldn't put it even into related_rules.

alexhaydock replied:
Well spotted. I've fixed this one and gone through the file with grep to check for any other instances of controls which are marked as Manual but which we have still set to `automated: yes`. There were none other than this one.

vojtapolasek reviewed (dismissed):
Found one thing which should be fixed.

matejak commented:
There are legitimate failures in the gating test, but it's just some CCEs and references. @alexhaydock do you think that you could look into it? An automated CCE adder can be extracted from #7249, and adding references will be very easy for you.
Also would it make sense to rebase the PR as #7226 got merged?

alexhaydock commented:
> There are [legitimate failures](https://github.com/ComplianceAsCode/content/pull/6976/checks?check_run_id=3299258805) in the gating test, but it's just some CCEs and references. @alexhaydock do you think that you could look into it? An automated CCE adder can be extracted from #7249, and adding references will be very easy for you.
> Also would it make sense to rebase the PR as #7226 got merged?

Thanks @matejak 

I think I've addressed the failures in the build now. I guess we will find out when the next run completes.

I did try the tool in #7249 and even after running `source .pyenv.sh` to set up the Python environment properly, I found that the script would run but didn't produce any output, and a `git diff` in the directory suggests it wasn't actually doing anything. I didn't really have time to debug though sadly.

This is what I was trying to run, inside the root dir of 8c8d6c7242edfdb00d47ee704f484eb35ce82910 and with the latest `fix-rules.py` from your branch:
```sh
~/fixrules/fix-rules.py -p products/rhel8/product.yml add-cce root_path_no_dot file_groupowner_efi_grub2_cfg file_owner_efi_grub2_cfg file_permissions_efi_grub2_cfg
```

matejak commented on linux_os/guide/system/accounts/accounts-restrictions/root_logins/use_pam_wheel_for_su/rule.yml:
```diff
@@ -24,6 +24,7 @@ identifiers:
 
 references:
     cis@rhel7: "5.7"
+    cis@rhel8: 5.7
```
I am sorry, but you have to quote the `5.7` for the time being, because otherwise it is interpreted as a number, which confuses the build system.

alexhaydock replied:
@matejak Ah that makes sense! I guess the references with multiple `.` delimiters are treated as strings automatically?

matejak commented:
Nice! And one more thing - if you rebase against master (i.e. you run `git rebase master`), those two merge commits should simply fall out. You will then have to `push --force-with-lease` after the rebase, but it is worth to get rid of those merge commits - we try to keep those out of the history.

pep8speaks commented:
Hello @alexhaydock! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`ssg/controls.py`](https://github.com/ComplianceAsCode/content/blob/9dae4cb4e1b9bb7a5bf1c7e2fc43eb722e10781d/ssg/controls.py):

> [Line 162:42](https://github.com/ComplianceAsCode/content/blob/9dae4cb4e1b9bb7a5bf1c7e2fc43eb722e10781d/ssg/controls.py#L162): [E741](https://duckduckgo.com/?q=pep8%20E741) ambiguous variable name 'l'
> [Line 162:100](https://github.com/ComplianceAsCode/content/blob/9dae4cb4e1b9bb7a5bf1c7e2fc43eb722e10781d/ssg/controls.py#L162): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (115 > 99 characters)
> [Line 163:21](https://github.com/ComplianceAsCode/content/blob/9dae4cb4e1b9bb7a5bf1c7e2fc43eb722e10781d/ssg/controls.py#L163): [E741](https://duckduckgo.com/?q=pep8%20E741) ambiguous variable name 'l'

alexhaydock commented:
@matejak ...I don't think those merge commands went quite as expected. Uh oh :grimacing: (or is that what you were expecting?)

matejak commented:
No, I think that everything went fine. Those failed actions indicate that an OVAL may be broken, but it probably doesn't have anything to do with the rebase - I have checked the branch before and after, and it looks the same.

alexhaydock commented:
> No, I think that everything went fine. Those failed actions indicate that an OVAL may be broken, but it probably doesn't have anything to do with the rebase - I have checked the branch before and after, and it looks the same.

I guess the GitHub UI just presents it in a confusing way then. It went from the "Files Changed" tab showing just my 5 changed files to showing 149 changed files, mostly from commits other than mine. But if that's expected then no problem!

I'll have a look through the build output when they've all finished running and see whether I can track down the broken OVAL.

cipherboy commented:
@matejak I agree with @alexhaydock, something looks funny here. Did you perhaps do the rebase/merge the wrong direction, and pulled stuff from master into this branch, rather than rebasing your branch on top of master? Just a thought.

matejak commented:
Alex probably rebased against an old master. File contents are fine, but when I have tried to rebase, I got just over 50 commits. So Alex please sync your master with upstream (i.e. this repository - `git pull` should do the trick, `git log` will give you a solid indication whether you were really successful with the update), and the rebase again. The number of commits and files changed should drop, and there should be no intervention required.

alexhaydock commented:
> @matejak I agree with @alexhaydock, something looks funny here. Did you (Alex) perhaps do the rebase/merge the wrong direction, and pulled stuff from master into this branch, rather than rebasing your branch on top of master? Just a thought.

@cipherboy 

Ah! I think you're absolutely right :facepalm: I always manage to mess up merges somehow. Thank goodness for `git reflog`. I've undone what I did for now and I'll try again later with a proper rebase.

matejak commented:
That's it, thank you very much for the effort!

matejak reviewed (approved):
The issue in the control file that Vojta discovered before has been fixed, and the PR has been cleanly rebased, so there is nothing else that stops it from merging.

matejak commented:
/retest

alexhaydock commented:
Thanks for the help and feedback everyone! Glad to finally get this one done and merged. :smile:

alexhaydock commented:
Out of interest @matejak (completely understand if you can't say) but do you know whether it's likely that CAC 0.1.58 will make it into RHEL 8.5, or is it more likely to be the 0.1.57 release that's already out?

matejak commented:
> Out of interest @matejak (completely understand if you can't say) but do you know whether it's likely that CAC 0.1.58 will make it into RHEL 8.5, or is it more likely to be the 0.1.57 release that's already out?

Unfortunately we can't comment on our plans, but you can try to read between the lines, and see how it went when 8.5 is out :-)
